modules/gapi/src/backends/oak/goakbackend.cpp

```diff
@@ -0,0 +1,616 @@
+    // The only heterogeneous case possible is if we insert other backends after or before
+    // OAK island.
+    std::unique_ptr<dai::Device> m_device;
+    std::unique_ptr<dai::Pipeline> m_pipeline;
+
+    std::vector<cv::GArg> m_oak_wrapped_args;
+
+public:
+    GOAKExecutable(const ade::Graph& g,
+                   const cv::GCompileArgs& args,
+                   const std::vector<ade::NodeHandle>& nodes,
+                   const std::vector<cv::gimpl::Data>& ins_data,
+                   const std::vector<cv::gimpl::Data>& outs_data);
+    ~GOAKExecutable() = default;
+
+    // FIXME: could it reshape?
+    virtual bool canReshape() const override { return false; }
+    virtual void reshape(ade::Graph&, const GCompileArgs&) override {
+        util::throw_error
+            (std::logic_error
+             ("GOAKExecutable::reshape() is not supported"));
+    }
+
+    virtual void handleNewStream() override;
+    virtual void handleStopStream() override;
+};
+
+class GOAKContext {
+public:
+    // FIXME: make private?
+    using Input = GOAKExecutable::ExtractTypeHelper::Input;
+    using Output = GOAKExecutable::ExtractTypeHelper::Output;
+    using InputPtr = GOAKExecutable::ExtractTypeHelper::Input*;
+    using OutputPtr = GOAKExecutable::ExtractTypeHelper::Output*;
+
+    GOAKContext(const std::unique_ptr<dai::Pipeline>& pipeline,
+                const std::tuple<int, int>& camera_size,
+                std::vector<cv::GArg>& args,
+                std::vector<OutputPtr>& results);
+
+    // Generic accessor API
+    template<typename T>
+    T& inArg(int input) { return m_args.at(input).get<T>(); }
+
+    // FIXME: consider not using raw pointers
+    InputPtr* in(int input);
+    OutputPtr* out(int output);
+
+    const std::unique_ptr<dai::Pipeline>& pipeline();
+    const std::tuple<int, int>& camera_size() const;
+
+private:
+    const std::unique_ptr<dai::Pipeline>& m_pipeline;
+    const std::tuple<int, int>& m_camera_size;
+    std::vector<cv::GArg>& m_args;
+    std::vector<OutputPtr>& m_outputs;
+};
+
+GOAKContext::GOAKContext(const std::unique_ptr<dai::Pipeline>& pipeline,
+                         const std::tuple<int, int>& camera_size,
+                         std::vector<cv::GArg>& args,
+                         std::vector<OutputPtr>& results)
+    : m_pipeline(pipeline), m_camera_size(camera_size), m_args(args), m_outputs(results) {}
+
+const std::unique_ptr<dai::Pipeline>& GOAKContext::pipeline() {
+    return m_pipeline;
+}
+
+const std::tuple<int, int>& GOAKContext::camera_size() const {
+    return m_camera_size;
+}
+
+GOAKContext::InputPtr* GOAKContext::in(int input) {
+    return inArg<GOAKContext::InputPtr*>(input);
+}
+
+GOAKContext::OutputPtr* GOAKContext::out(int output) {
+    return &(m_outputs.at(output));
+}
+
+namespace detail {
+template<class T> struct get_in;
+template<> struct get_in<cv::GFrame> {
+    static GOAKContext::InputPtr* get(GOAKContext &ctx, int idx) { return ctx.in(idx); }
```

Comment:
so why a pointer to pointer?

Reply:
Since we are changing inputs and outputs individually in kernels, those need to be `**`. Do you suggest we start storing and copying actual `dai::Input` and `dai::Output` instead of pointers?